src/engine/CodeValidator.v3

```diff
@@ -1318,10 +1318,16 @@ class CodeValidator(extensions: Extension.set, limits: Limits, module: Module, e
 		var p = labelArgs(target);
 		if (p == null) return;
 		var ct = if(c.tag == null, SigCache.arr_v, c.tag.fields);
-		if (c.exnref) ct = Arrays.append(ValueTypes.EXNREF, ct); // XXX: avoid array copy
-		if (ct.length != p.length) return err_atpc().ArityMismatchInHandler("catch", i, p.length, ct.length);
+		
+		// If `c.exnref` is true, then `ValueTypes.EXNREF` is appended. This is
+		// simulated without an extra array copy of the signatures.
+		var actual_ct_length = if(c.exnref, ct.length + 1, ct.length);
+		if (actual_ct_length != p.length) {
+			return err_atpc().ArityMismatchInHandler("catch", i, p.length, actual_ct_length);
+		}
 		for (j < p.length) {
```

Comment:
I think this will be a little more efficient to factor the branch out of the loop; i.e. to do a check against the `ct` array first, then optionally check the `exnref`.